### PR TITLE
Enhance Qt Crash Course Documentation with Detailed Mouse and Keyboard Event Handling

### PR DESCRIPTION
**Pull request by Lamis-Abla**
Added detailed examples for handling keyboard and mouse events in PyQt applications. These additions include code snippets and explanations on how to effectively implement `keyPressEvent` and `mousePressEvent` within the PyQt framework.

Please review the changes in the following file: [qtcrashcourse.rst](doc/source/getting_started/qtcrashcourse.rst)

### CODE REVIEW

File: testfile.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+"Dies ist eine Testdatei"
```
Comment (j9ac9k) [inline testfile.txt:1]: We should probably remove this file :)

File: doc/source/getting_started/qtcrashcourse.rst [hunk 2]
```diff
@@ -91,7 +91,93 @@ http://qt-project.org/doc/qt-4.8/graphicsview.html#the-graphics-view-coordinate-
...
 
+- Events in Qt are conceptualized as user interactions with the application, each represented by an event object (QEvent).
+- Various types of events correspond to different user interactions.
+- Event objects encapsulate details concerning the specific occurrence.
+- Dispatched to designated event handlers within the widget where the interaction occurs, these events allow for customizable responses.
+- Handlers may be extended or redefined to modify widget response to interactions.
+
+Mouse Events
+^^^^^^^^^^^^
+Interactions such as clicks, movements, and button releases are managed by overriding methods including ``mousePressEvent``, ``mouseReleaseEvent``, ``mouseDoubleClickEvent``, and ``mouseMoveEvent``.
```
Comment (j9ac9k) [inline doc/source/getting_started/qtcrashcourse.rst:106]: These methods should probably link to their upstream methods.   I think the syntax to get the linking to the upstream docs is:

```
:meth:`QtWidget.mousePressEvent <PyQt6.QtWidgets.QtWidget.mousePressEvent>`
:meth:`QtWidget.mouseReleaseEvent <PyQt6.QtWidgets.QtWidget.mouseReleaseEvent>`
:meth:`QtWidget.mouseDoubleClickEvent <PyQt6.QtWidgets.QtWidget.mouseDoubleClickEvent>`
:meth:`QtWidget.mouseMoveEvent <PyQt6.QtWidgets.QtWidget.mouseMoveEvent>`
```

File: doc/source/getting_started/qtcrashcourse.rst [hunk 3]
```diff
@@ -91,7 +91,93 @@ http://qt-project.org/doc/qt-4.8/graphicsview.html#the-graphics-view-coordinate-
...
+    from PyQt6.QtWidgets import QApplication, QMainWindow # Should work with PyQt5 / PySide2 / PySide6 as well
+    from PyQt6.QtCore import Qt
+
+    class MainWindow(QMainWindow):
+        def __init__(self):
+            super().__init__()
+            self.setWindowTitle('Mouse and Keyboard Event Demo') # Sets the Title of the window
+            self.setGeometry(100, 100, 400, 300) # Sets the position and size of the window
+
+        def mousePressEvent(self, event): # This method checks if the left mouse button was pressed on the widget and prints the position of the click.
```
Comment (j9ac9k) [inline doc/source/getting_started/qtcrashcourse.rst:123]: There is a horizontal scrolling bar, try and keep the line-width here really short. In the rendered docs it looks like:

image

**Comment by j9ac9k** [comment]
Thank you so much for this PR @Lamis-Abla 

I have a few nit-pick items, such as replacing references to various upstream methods with links that will go to their respective documentation and narrowing the line length (I personally set a ruler in my IDE for 88 characters and try not to go past that).

The content of the documentation looks great!

File: doc/source/getting_started/qtcrashcourse.rst (hunk 3 above)
Comment (Lamis-Abla) [reply doc/source/getting_started/qtcrashcourse.rst:123]: Hi [j9ac9k](https://github.com/j9ac9k),

Thank you for your feedback!

I wanted to clarify your recommendation about the line length. Does the 88-character limit apply only to the code lines, or should I also adjust the text in the documentation to adhere to this limit?

Thanks again for your guidance!
Lamis

**Comment by j9ac9k** [comment]
I try to conform the 88 character line length to everything is rst files.
Sphinx will ignore single newline characters so adding new lines does not
impact the rendered result, it just makes it easier to work with in a text
editor.

I’m sure there are places where this line limit won’t make sense (like
linking to something that had a long link) so there is room for exceptions.

File: doc/source/getting_started/qtcrashcourse.rst (hunk 3 above)
Comment (j9ac9k) [reply doc/source/getting_started/qtcrashcourse.rst]: I try to conform the 88 character line length to everything is rst files.
Sphinx will ignore single newline characters so adding new lines does not
impact the rendered result, it just makes it easier to work with in a text
editor.

I’m sure there are places where this line limit won’t make sense (like
linking to something that had a long link) so there is room for exceptions.

File: doc/source/getting_started/qtcrashcourse.rst [hunk 4]
```diff
@@ -91,7 +91,131 @@ http://qt-project.org/doc/qt-4.8/graphicsview.html#the-graphics-view-coordinate-
...
+- Events in Qt are conceptualized as user interactions with the application, each represented by an event object (QEvent).
+- Various types of events correspond to different user interactions.
+- Event objects encapsulate details concerning the specific occurrence.
+- Dispatched to designated event handlers within the widget where the interaction occurs, these events allow for customizable responses.
+- Handlers may be extended or redefined to modify widget response to interactions.
+
+Mouse Events
+^^^^^^^^^^^^
+Mouse Events
+^^^^^^^^^^^^
```
Comment (j9ac9k) [inline doc/source/getting_started/qtcrashcourse.rst:107]: Probably don't need 2x the section headers :)

File: doc/source/getting_started/qtcrashcourse.rst [hunk 5]
```diff
@@ -91,7 +91,131 @@ http://qt-project.org/doc/qt-4.8/graphicsview.html#the-graphics-view-coordinate-
...
+Interactions such as clicks, movements, and button releases are managed by
+overriding methods including  mousePressEvent_ ,
```
Comment (j9ac9k) [inline doc/source/getting_started/qtcrashcourse.rst:109]: Does the sphinx-cross-linking not work here?  I would imagine the following would work? 

```rst
:meth:`QGraphicsItem.mousePressEvent <QGraphicsItem.mousePressEvent>`
```

If not let me know and I'll look into it more.

File: doc/source/getting_started/qtcrashcourse.rst [hunk 6]
```diff
@@ -91,7 +91,131 @@ http://qt-project.org/doc/qt-4.8/graphicsview.html#the-graphics-view-coordinate-
...
+    # and prints the position of the click.
+    if event.button() == Qt.MouseButton.LeftButton:
+    print("Left mouse button pressed at:", event.position())
+
+    # Initialize the QApplication
+    app = QApplication([])
+    window = MainWindow()
+    window.show()
+    # Start the event loop
+    app.exec()
```
Comment (j9ac9k) [inline doc/source/getting_started/qtcrashcourse.rst:149]: Looks like indentation got broken ?

File: doc/source/getting_started/qtcrashcourse.rst [hunk 7]
```diff
@@ -91,7 +91,131 @@ http://qt-project.org/doc/qt-4.8/graphicsview.html#the-graphics-view-coordinate-
...
+    # The 'event.text()' method retrieves the character or characters 
+    # associated with the key press, and then prints it to the console.
+    print(f"Key pressed: {event.text()}")
+
+    # Initialize the QApplication
+    app = QApplication([])
+    window = MainWindow()
+    window.show()
+    # Start the event loop
+    app.exec()
```
Comment (j9ac9k) [inline doc/source/getting_started/qtcrashcourse.rst:198]: Again, looks like indentation got broken in the example code..

File: doc/source/getting_started/qtcrashcourse.rst [hunk 8]
```diff
@@ -91,6 +91,142 @@ http://qt-project.org/doc/qt-4.8/graphicsview.html#the-graphics-view-coordinate-
...
+the event is either handled or reaches the main window.
+
+It is facilitated by methods such as 
+`accept`_, and `ignore`_, which allow developers to 
+exert precise control over the event flow.
+
+.. _accept: https://doc.qt.io/qt-6/qevent.html#accept
+.. _ignore: https://doc.qt.io/qt-6/qevent.html#ignore
+
+Example: Custom Event Handling ::
```
Comment (j9ac9k) [inline doc/source/getting_started/qtcrashcourse.rst:223]: This block is  still long enough that's being clipped in the rendered output:

image

**Comment by j9ac9k** [comment]
superseded by #3087

**Pull request closed without merging**